cde/lib/tt/man/man5/tttk.5
.de LI
.\" simulate -mm .LIs by turning them into .TPs
.TP \\n()Jn
\\$1
..
.TH tttk 5 "1 March 1996" "ToolTalk 1.3" "ToolTalk Definitions"
.BH "1 March 1996"
.\" CDE Common Source Format, Version 1.0.0
.\" (c) Copyright 1993, 1994 Hewlett-Packard Company
.\" (c) Copyright 1993, 1994 International Business Machines Corp.
.\" (c) Copyright 1993, 1994 Sun Microsystems, Inc.
.\" (c) Copyright 1993, 1994 Novell, Inc.
.IX "Tt/tttk.h" "" "Tt/tttk.h \(em ToolTalk definitions" ""
.SH "NAME"
tttk, Tttttk \- ToolTalk definitions
.SH "SYNOPSIS"
.ft 3
.nf
#include <Tt/tttk.h>
.PP
.fi
.SH "DESCRIPTION"
.PP
The
.BR <Tt/tttk.h> 
header defines the following enumeration data type,
with at least the following members:
.PP
.RS 3
.nr )J 6
.na
.LI \f3Tttk_op\fP
.br
.BR \s-1TTDT_CREATED\s+1 ,
.BR \s-1TTDT_DELETED\s+1 ,
.BR \s-1TTDT_DO_COMMAND\s+1 ,
.BR \s-1TTDT_GET_ENVIRONMENT\s+1 ,
.BR \s-1TTDT_GET_GEOMETRY\s+1 ,
.BR \s-1TTDT_GET_ICONIFIED\s+1 ,
.BR \s-1TTDT_GET_LOCALE\s+1 ,
.BR \s-1TTDT_GET_MAPPED\s+1 ,
.BR \s-1TTDT_GET_MODIFIED\s+1 ,
.BR \s-1TTDT_GET_SITUATION\s+1 ,
.BR \s-1TTDT_GET_STATUS\s+1 ,
.BR \s-1TTDT_GET_SYSINFO\s+1 ,
.BR \s-1TTDT_GET_XINFO\s+1 ,
.BR \s-1TTDT_LOWER\s+1 ,
.BR \s-1TTDT_MODIFIED\s+1 ,
.BR \s-1TTDT_MOVED\s+1 ,
.BR \s-1TTDT_OP_LAST\s+1 ,
.BR \s-1TTDT_OP_NONE\s+1 ,
.BR \s-1TTDT_PAUSE\s+1 ,
.BR \s-1TTDT_QUIT\s+1 ,
.BR \s-1TTDT_RAISE\s+1 ,
.BR \s-1TTDT_RESUME\s+1 ,
.BR \s-1TTDT_REVERT\s+1 ,
.BR \s-1TTDT_REVERTED\s+1 ,
.BR \s-1TTDT_SAVE\s+1 ,
.BR \s-1TTDT_SAVED\s+1 ,
.BR \s-1TTDT_SET_ENVIRONMENT\s+1 ,
.BR \s-1TTDT_SET_GEOMETRY\s+1 ,
.BR \s-1TTDT_SET_ICONIFIED\s+1 ,
.BR \s-1TTDT_SET_LOCALE\s+1 ,
.BR \s-1TTDT_SET_MAPPED\s+1 ,
.BR \s-1TTDT_SET_SITUATION\s+1 ,
.BR \s-1TTDT_SET_XINFO\s+1 ,
.BR \s-1TTDT_SIGNAL\s+1 ,
.BR \s-1TTDT_STARTED\s+1 ,
.BR \s-1TTDT_STATUS\s+1 ,
.BR \s-1TTDT_STOPPED\s+1 ,
.BR \s-1TTME_ABSTRACT\s+1 ,
.BR \s-1TTME_COMPOSE\s+1 ,
.BR \s-1TTME_DEPOSIT\s+1 ,
.BR \s-1TTME_DISPLAY\s+1 ,
.BR \s-1TTME_EDIT\s+1 ,
.BR \s-1TTME_INTERPRET\s+1 ,
.BR \s-1TTME_MAIL\s+1 ,
.BR \s-1TTME_MAIL_COMPOSE\s+1 ,
.BR \s-1TTME_MAIL_EDIT\s+1 ,
.BR \s-1TTME_PRINT\s+1 ,
.BR \s-1TTME_TRANSLATE\s+1
.PP
.RE
.nr )J 0
.PP
.ad
The header declares the following global
string constants for some standard vtypes:
.PP
.ft 3
.nf
extern const char  *Tttk_boolean:
extern const char  *Tttk_file:
extern const char  *Tttk_height:
extern const char  *Tttk_integer:
extern const char  *Tttk_message_id:
extern const char  *Tttk_string:
extern const char  *Tttk_title:
extern const char  *Tttk_width:
extern const char  *Tttk_xoffset:
extern const char  *Tttk_yoffset:
.PP
.fi
.PP
The header declares the following as functions:
.PP
.ft 3
.nf
.ta \w'int ttdt_Get_Modified('u
int ttdt_Get_Modified(Tt_message \f2context\fP,
	const char *\f2pathname\fP,
	Tt_scope \f2the_scope\fP,
	XtAppContext \f2app2run\fP,
	int \f2ms_timeout\fP);
.PP
.fi
.PP
.ft 3
.nf
.ta \w'Tt_status ttdt_Revert('u
Tt_status ttdt_Revert(Tt_message \f2context\fP,
	const char *\f2pathname\fP,
	Tt_scope \f2the_scope\fP,
	XtAppContext \f2app2run\fP,
	int \f2ms_timeout\fP);
.PP
.fi
.PP
.ft 3
.nf
.ta \w'Tt_status ttdt_Save('u
Tt_status ttdt_Save(Tt_message \f2context\fP,
	const char *\f2pathname\fP,
	Tt_scope \f2the_scope\fP,
	XtAppContext \f2app2run\fP,
	int \f2ms_timeout\fP);
.PP
.fi
.PP
.ft 3
.nf
.ta \w'Tt_status ttdt_close('u
Tt_status ttdt_close(const char *\f2procid\fP,
	const char *\f2new_procid\fP,
	int \f2sendStopped\fP);
.PP
.fi
.PP
.ft 3
.nf
.ta \w'Tt_status ttdt_file_event('u
Tt_status ttdt_file_event(Tt_message \f2context\fP,
	Tttk_op \f2event\fP,
	Tt_pattern *\f2patterns\fP,
	int \f2send\fP);
.PP
.fi
.PP
.ft 3
.nf
.ta \w'Tt_pattern *ttdt_file_join('u
Tt_pattern *ttdt_file_join(const char *\f2pathname\fP,
	Tt_scope \f2the_scope\fP,
	int \f2join\fP,
	Ttdt_file_cb \f2cb\fP,
	void *\f2clientdata\fP);
.PP
.fi
.PP
.ft 3
.nf
.ta \w'Tt_message ttdt_file_notice('u
Tt_message ttdt_file_notice(Tt_message \f2context\fP,
	Tttk_op \f2op\fP,
	Tt_scope \f2scope\fP,
	const char *\f2pathname\fP,
	int \f2send_and_destroy\fP);
.PP
.fi
.PP
.ft 3
.nf
.ta \w'Tt_status ttdt_file_quit('u
Tt_status ttdt_file_quit(Tt_pattern *\f2patterns\fP,
	int \f2quit\fP);
.PP
.fi
.PP
.ft 3
.nf
.ta \w'Tt_message ttdt_file_request('u
Tt_message ttdt_file_request(Tt_message \f2context\fP,
	Tttk_op \f2op\fP,
	Tt_scope \f2scope\fP,
	const char *\f2pathname\fP,
	Ttdt_file_cb \f2cb\fP,
	void *\f2client_data\fP,
	int \f2send_and_destroy\fP);
.PP
.fi
.PP
.ft 3
.nf
.ta \w'Tt_pattern *ttdt_message_accept('u
Tt_pattern *ttdt_message_accept(Tt_message \f2contract\fP,
	Ttdt_contract_cb \f2cb\fP,
	void *\f2clientdata\fP,
	Widget \f2shell\fP,
	int \f2accept\fP,
	int \f2sendStatus\fP);
.PP
.fi
.PP
.ft 3
.nf
.ta \w'char *ttdt_open('u
char *ttdt_open(int *\f2ttfd\fP,
	const char *\f2toolname\fP,
	const char *\f2vendor\fP,
	const char *\f2version\fP,
	int \f2sendStarted\fP);
.PP
.fi
.PP
.ft 3
.nf
.ta \w'Tt_status ttdt_sender_imprint_on('u
Tt_status ttdt_sender_imprint_on(const char *\f2handler\fP,
	Tt_message \f2contract\fP,
	char **\f2display\fP,
	int *\f2width\fP,
	int *\f2height\fP,
	int *\f2xoffset\fP,
	int *\f2yoffset\fP,
	XtAppContext \f2app2run\fP,
	int \f2ms_timeout\fP);
.PP
.fi
.PP
.ft 3
.nf
.ta \w'Tt_pattern *ttdt_session_join('u
Tt_pattern *ttdt_session_join(const char *\f2sessid\fP,
	Ttdt_contract_cb \f2cb\fP,
	Widget \f2shell\fP,
	void *\f2clientdata\fP,
	int \f2join\fP);
.PP
.fi
.PP
.ft 3
.nf
.ta \w'Tt_status ttdt_session_quit('u
Tt_status ttdt_session_quit(const char *\f2sessid\fP,
	Tt_pattern *\f2sess_pats\fP,
	int \f2quit\fP);
.PP
.fi
.PP
.ft 3
.nf
.ta \w'Tt_pattern *ttdt_subcontract_manage('u
Tt_pattern *ttdt_subcontract_manage(Tt_message \f2subcontract\fP,
	Ttdt_contract_cb \f2cb\fP,
	Widget \f2shell\fP,
	void *\f2clientdata\fP);
.PP
.fi
.PP
.ft 3
.nf
.ta \w'Tt_status ttmedia_Deposit('u
Tt_status ttmedia_Deposit(Tt_message \f2load_contract\fP,
	const char *\f2buffer_id\fP,
	const char *\f2media_type\fP,
	const unsigned char *\f2new_contents\fP,
	int \f2new_len\fP,
	const char *\f2pathname\fP,
	XtAppContext \f2app2run\fP,
	int \f2ms_timeout\fP);
.PP
.fi
.PP
.ft 3
.nf
.ta \w'Tt_message ttmedia_load('u
Tt_message ttmedia_load(Tt_message \f2context\fP,
	Ttmedia_load_msg_cb \f2cb\fP,
	void *\f2clientdata\fP,
	Tttk_op \f2op\fP,
	const char *\f2media_type\fP,
	const unsigned char *\f2contents\fP,
	int \f2len\fP,
	const char *\f2file\fP,
	const char *\f2docname\fP,
	int \f2send\fP);
.PP
.fi
.PP
.ft 3
.nf
.ta \w'Tt_message ttmedia_load_reply('u
Tt_message ttmedia_load_reply(Tt_message \f2contract\fP,
	const unsigned char *\f2new_contents\fP,
	int \f2new_len\fP,
	int \f2reply_and_destroy\fP);
.PP
.fi
.PP
.ft 3
.nf
.ta \w'Tt_status ttmedia_ptype_declare('u
Tt_status ttmedia_ptype_declare(const char *\f2ptype\fP,
	int \f2base_opnum\fP,
	Ttmedia_load_pat_cb \f2cb\fP,
	void *\f2clientdata\fP,
	int \f2declare\fP);
.PP
.fi
.PP
.ft 3
.nf
.ta \w'void tttk_Xt_input_handler('u
void tttk_Xt_input_handler(XtPointer \f2procid\fP,
	int *\f2source\fP,
	XtInputId *\f2id\fP);
.PP
.fi
.PP
.ft 3
.nf
.ta \w'Tt_status tttk_block_while('u
Tt_status tttk_block_while(XtAppContext \f2app2run\fP,
	const int *\f2blocked\fP,
	int \f2ms_timeout\fP);
.PP
.fi
.PP
.ft 3
.nf
.ta \w'Tt_status tttk_message_abandon('u
Tt_status tttk_message_abandon(Tt_message \f2msg\fP);
.PP
.fi
.PP
.ft 3
.nf
.ta \w'Tt_message tttk_message_create('u
Tt_message tttk_message_create(Tt_message \f2context\fP,
	Tt_class \f2the_class\fP,
	Tt_scope \f2the_scope\fP,
	const char *\f2handler\fP,
	const char *\f2op\fP,
	Tt_message_callback \f2callback\fP);
.PP
.fi
.PP
.ft 3
.nf
.ta \w'Tt_status tttk_message_destroy('u
Tt_status tttk_message_destroy(Tt_message \f2msg\fP);
.PP
.fi
.PP
.ft 3
.nf
.ta \w'Tt_status tttk_message_fail('u
Tt_status tttk_message_fail(Tt_message \f2msg\fP,
	Tt_status \f2status\fP,
	const char *\f2status_string\fP,
	int \f2destroy\fP);
.PP
.fi
.PP
.ft 3
.nf
.ta \w'Tt_status tttk_message_reject('u
Tt_status tttk_message_reject(Tt_message \f2msg\fP,
	Tt_status \f2status\fP,
	const char *\f2status_string\fP,
	int \f2destroy\fP);
.PP
.fi
.PP
.ft 3
.nf
.ta \w'char *tttk_op_string('u
char *tttk_op_string(Tttk_op \f2opcode\fP);
.PP
.fi
.PP
.ft 3
.nf
.ta \w'Tttk_op tttk_string_op('u
Tttk_op tttk_string_op(const char *\f2opstring\fP);
.PP
.fi


cde/lib/tt/lib/tttk/tttk.C
//%%  (c) Copyright 1993, 1994 Hewlett-Packard Company			
//%%  (c) Copyright 1993, 1994 International Business Machines Corp.	
//%%  (c) Copyright 1993, 1994 Sun Microsystems, Inc.			
//%%  (c) Copyright 1993, 1994 Novell, Inc. 				
//%%  $Revision: 1.13 $ 			 				
/*
 * @(#)tttk.C	1.3	93/09/07
 *
 * Copyright 1993 Sun Microsystems, Inc.  All rights reserved.
 */
#include <poll.h>
#include <errno.h>
#include <unistd.h>
#include <limits.h>
#include "api/c/tt_c.h"
#include "util/tt_gettext.h"
#include "util/tt_Xlib.h"
#include "util/tt_port.h"
#include "tttk/tttk.h"
#include "tttk/tttk2free.h"
#include "tttk/tttkutils.h"
#include "tttk/tttkpattern.h"
#include "tttk/ttdtprocid.h"
#include "tttk/ttdesktop.h"

extern Tt_status	_tt_errno_status( int err_no );

const char  *Tttk_integer		= "integer";
const char  *Tttk_string		= "string";
const char  *Tttk_boolean		= "boolean";
const char  *Tttk_file			= "File";
const char  *Tttk_message_id		= "messageID";
const char  *Tttk_title			= "title";
const char  *Tttk_width			= "width";
const char  *Tttk_height		= "height";
const char  *Tttk_xoffset		= "xOffset";
const char  *Tttk_yoffset		= "yOffset";

_TtDtProcid *_ttdtme			= 0;

// API calls below...

char *
ttdt_open(
	int	       *ttFD,
	const char     *toolName,
	const char     *vendor,
	const char     *version,
	int		sendStarted
)
{
	_ttdtme = new _TtDtProcid( toolName, vendor, version );
	if (_ttdtme == 0) {
		return (char *)tt_error_pointer( TT_ERR_NOMEM );
	}
	char *procID = tt_open();
	Tt_status status = tt_ptr_error( *procID );
	if (status != TT_OK) {
		return procID;
	}
	*ttFD = tt_fd();
	status = tt_int_error( *ttFD );
	if (status != TT_OK) {
		tt_free( procID );
		tt_close();
		return (char *)tt_error_pointer( status );
	}
	if (sendStarted) {
		ttdt_Started( 0, toolName, vendor, version, 1 );
	}
	return procID;
}

//
// A callback used internally to pick geometry info out of
// a Get_Geometry reply and stuff it into (DisplayInfo *)clientData.
//
static Tt_message
_ttDtApplyGeom(
	Tt_message	msg,
	void	       *clientData,
	Tt_message	,
	int		width,
	int		height,
	int		xOffset,
	int		yOffset
)
{
	DisplayInfo *info = (DisplayInfo *)clientData;
	if (info != 0) {
		info->repliesOutStanding--;
		info->width = width;
		info->height = height;
		info->xOffset = xOffset;
		info->yOffset = yOffset;
	}
	tttk_message_destroy( msg );
	return 0;
}

//
// A TtDtGetXInfoMsgCB used internally to pick X11 info out of
// a Get_XInfo reply and stuff it into (DisplayInfo *)clientData.
//
static Tt_message
_ttDtApplyXInfo(
	Tt_message	msg,
	void	       *clientData,
	Tt_message	,
	char           *display,
	int,		//visual
	int		//depth
)
{
	DisplayInfo *info = (DisplayInfo *)clientData;
	if (info != 0) {
		info->repliesOutStanding--;
		info->display = display;
	}
	tttk_message_destroy( msg );
	return 0;
}

Tt_status
ttdt_sender_imprint_on(
	const char     *handler,
	Tt_message	commission,
	char	      **pDisplay,
	int	       *width,
	int	       *height,
	int	       *xOffset,
	int	       *yOffset,
	XtAppContext	app2run,
	int		msTimeOut
)
{
	Tt_status status;
	Tt_message msg;
	DisplayInfo info;
	info.repliesOutStanding = 0;
	info.width = -1;
	info.height = -1;
	info.xOffset = INT_MAX;
	info.yOffset = INT_MAX;
	info.display = 0;
	if ((width != 0) || (height != 0) || (xOffset != 0) || (yOffset != 0)){
		msg = ttdt_Get_Geometry( handler, commission,
					 _ttDtApplyGeom, &info, 1 );
		status = tt_ptr_error( msg );
		if (status == TT_OK) {
			info.repliesOutStanding++;
		}
	}
	// XXX sending XInfo before Geometry causes segv in handler!? Saberize.
	msg = ttdt_Get_XInfo( handler, commission,
			      _ttDtApplyXInfo, &info, 1);
	status = tt_ptr_error( msg );
	if (status == TT_OK) {
		info.repliesOutStanding++;
	}
	msg = ttdt_Get_Locale( handler, commission, 0, &info, 0, 1 );
	status = tt_ptr_error( msg );
	if (status == TT_OK) {
		info.repliesOutStanding++;
	}
	msg = ttdt_Get_Situation( handler, commission, 0, &info, 1 );
	status = tt_ptr_error( msg );
	if (status == TT_OK) {
		info.repliesOutStanding++;
	}
	if (info.repliesOutStanding == 0) {
		return status;
	}
	status = tttk_block_while( app2run,
				   &info.repliesOutStanding, msTimeOut );
	if (pDisplay == 0) {
		if (info.display != 0) {
			_tt_putenv( "DISPLAY", info.display );
		}
		tt_free( info.display );
	} else {
		*pDisplay = info.display;
	}
	if (width != 0) {
		*width = info.width;
	}
	if (height != 0) {
		*height = info.height;
	}
	if (xOffset != 0) {
		*xOffset = info.xOffset;
	}
	if (yOffset != 0) {
		*yOffset = info.yOffset;
	}
	return status;
}

Tt_status
ttdt_close(
	const char     *procid,
	const char     *newProcid,
	int		sendStopped
)
{
	Tt_status status;
	if (procid != 0) {
		status = tt_default_procid_set( procid );
		if (status != TT_OK) {
			return status;
		}
	}
	if (sendStopped) {
		ttdt_Stopped( 0, _ttdtme->toolname(), _ttdtme->vendor(),
			      _ttdtme->version(), 1 );
	}
	status = tt_close();
	if (status != TT_OK) {
		return status;
	}
	if (newProcid != 0) {
		status = tt_default_procid_set( newProcid );
		if (status != TT_OK) {
			return status;
		}
	}
	return TT_OK;
}

Tt_pattern *
ttdt_session_join(
	const char	       *sessid,
	Ttdt_contract_cb	cb,
	Widget			shell,
	void		       *clientdata,
	int			join
)
{
	Tt_status	status;
	_TttkItem2Free	item2free;
	if (sessid == 0) {
		sessid = tt_default_session();
		status = tt_ptr_error( sessid );
		if (status != TT_OK) {
			return (Tt_pattern *)tt_error_pointer( status );
		}
		item2free = (char *)sessid;
	}
	Tt_pattern *pats = _ttdtme->pats_create( 0, cb, shell, clientdata );
	if (join) {
		status = tt_session_join( sessid );
		if (status != TT_OK) {
			return (Tt_pattern *)tt_error_pointer( status );
		}
	}
	return pats;
}

Tt_status
ttdt_session_quit(
	const char     *sessid,
	Tt_pattern     *pats,
	int		quit
)
{
	Tt_status	status;
	_TttkItem2Free	item2free;
	if (sessid == 0) {
		sessid = tt_default_session();
		status = tt_ptr_error( sessid );
		if (status != TT_OK) {
			return status;
		}
		item2free = (char *)sessid;
	}
	status = _tttk_patterns_destroy( pats );
	if (status != TT_OK) {
		return status;
	}
	if (quit) {
		status = tt_session_quit( sessid );
		if (status != TT_OK) {
			return status;
		}
	}
	return status;
}

Tt_pattern *
ttdt_message_accept(
	Tt_message	contract,
	Ttdt_contract_cb cb,
	Widget		shell,
	void	       *clientData,
	int		accept,
	int		sendStatus
)
{
	Tt_status status;
	if (accept) {
		status = tt_message_accept( contract );
		if (status != TT_OK) {
			// Not fatal
		}
	}
	Tt_pattern *pats = _ttdtme->pats_create( contract, cb, shell,
						 clientData );
	status = tt_ptr_error( pats );
	if (status == TT_OK) {
		tt_message_user_set( contract, _TttkContractKey, pats );
	}
	if (sendStatus) {
		ttdt_Status( contract, contract,
			     catgets( _ttcatd, 1, 24, "Accepting request" ),
			     _ttdtme->toolname(), _ttdtme->vendor(),
			     _ttdtme->version(), 1 );
	}
	return pats;
}

Tt_pattern *
ttdt_subcontract_manage(
	Tt_message		subcontract,
	Ttdt_contract_cb	cb,
	Widget			shell,
	void		       *clientData
)
{
	const int numPats = 4;
	Tt_pattern *pats = (Tt_pattern *)tt_malloc(numPats * sizeof(Tt_pattern));
	if (pats == 0) {
		return (Tt_pattern *)tt_error_pointer( TT_ERR_NOMEM );
	}
	pats[ 0 ] = 0;
	if (shell != 0) {
		pats[0] = ttdt_Get_Geometry_pat( TT_HANDLE, subcontract,
				_ttdt_do_GSet_Geometry, shell, 1 );
		pats[1] = ttdt_Get_XInfo_pat( TT_HANDLE, subcontract,
				_ttdt_do_Get_XInfo, shell, 1 );
	} else {
		pats[0] = ttdt_Get_Geometry_pat( TT_HANDLE, subcontract,
				(Ttdt_Geometry_out_cb)cb, clientData, 1 );
		pats[1] = ttdt_Get_XInfo_pat(	TT_HANDLE, subcontract,
				(Ttdt_XInfo_out_cb)cb, clientData, 1 );
	}
	pats[2] = _ttdt_pat(	TTDT_STATUS, _ttdt_contract_cb, TT_OBSERVE,
				subcontract, cb, clientData, 1 );
	pats[ numPats - 1 ] = 0;
	for (int i = 0; i < numPats; i++) {
		Tt_status status = tt_ptr_error( pats[ i ] );
		if (status != TT_OK) {
			_tttk_patterns_destroy( pats );
			return (Tt_pattern *)tt_error_pointer( status );
		}
	}
	tt_message_user_set( subcontract, _TttkSubContractKey, pats );
	return pats;
}

void
tttk_Xt_input_handler(
	XtPointer  procid,
	int       *,
	XtInputId *id
)
{
	static const char *here = "ttdt_Xt_input_handler()";

	Tt_message msg = _tttk_message_receive( (char *)procid );
	Tt_status status = tt_ptr_error( msg );
	if (status != TT_OK) {
		if (status == TT_ERR_NOMP) {
			ttdt_close( (const char *)procid, 0, 0 );
			if (! _tt_load_xt()) {
				return;
			}
			if (id != 0) {
				CALLXT(XtRemoveInput)( *id );
			}
		}
		return;
	}
	if (msg == 0) {
		return;
	}
	(void) tttk_message_abandon( msg );
}

Tt_status
_tttk_block_procid_while(
	const int     *blocked,
	int            msTimeOut
)
{
	int fd = tt_fd();
	Tt_status status = tt_int_error( fd );
	if (status != TT_OK) {
		return status;
	}
	while ((blocked == 0) || (*blocked > 0)) {
		struct pollfd fds[ 1 ];
		fds[0].fd = fd;
		fds[0].events = POLLIN;
		int activeFDs = poll( fds, 1, msTimeOut );
		if (activeFDs == 0) {
			return TT_DESKTOP_ETIMEDOUT;
		} else if (activeFDs < 0) {
			if ((blocked != 0) && (*blocked)) {
				_ttDtPError( "_tttk_block_procid_while()",
					     "poll()" );
				return _tt_errno_status( errno );
			}
		}
		if (fds[ 0 ].revents != 0) {
			//
			// AIX has fd in struct pollfd as a long, not
			// an int, and complains (justifiably) when
			// long* is passed when int* is needed. So use
			// call-by-value-return instead of
			// call-by-reference.  Hey,
			// tttk_Xt_input_handler doesn\'t use the
			// value anyway...
			//
			int fd_temp;
			fd_temp = fds[0].fd;
			tttk_Xt_input_handler( 0, &fd_temp, 0 );
			fds[0].fd = fd_temp;
		}
		if (msTimeOut == 0) {
			return TT_OK;
		}
	}
	return TT_OK;
}

void
_tttk_timed_out(
	XtPointer	p_timed_out,
	XtIntervalId
)
{
	*(int *)p_timed_out = 1;
}

Tt_status
_tttk_block_app_while(
	XtAppContext	app2run,
	const int     *blocked,
	int            msTimeOut
)
{
	XtIntervalId alarm = 0;
	if ((blocked != 0) && (*blocked <= 0)) return TT_OK;
	int timed_out = (msTimeOut == 0);
	if ((! _tt_load_xt()) || (! _tt_load_xlib())) {
		return TT_ERR_ACCESS; // i.e. ELIBACC
	}
	if ((timed_out) && (! CALLXT(XtAppPending)( app2run ))) {
		//
		// We are non-blocking and no input is available,
		// so return.
		//
		return TT_DESKTOP_ETIMEDOUT;
	}
	if (msTimeOut > 0) {
		alarm = CALLXT(XtAppAddTimeOut)( app2run,
					(unsigned long)msTimeOut,
					(XtTimerCallbackProc)_tttk_timed_out,
					&timed_out );
	}
	do {
		CALLXT(XtAppProcessEvent)( app2run, XtIMAll );
	} while ((! timed_out) && ((blocked == 0) || (*blocked > 0)));
	if (alarm != 0) {
		CALLXT(XtRemoveTimeOut)( alarm );
	}
	if (timed_out && (msTimeOut > 0)) {
		return TT_DESKTOP_ETIMEDOUT;
	}
	return TT_OK;
}

Tt_status
tttk_block_while(
	XtAppContext	app2run,
	const int     *blocked,
	int            msTimeOut
)
{
	if (app2run == 0) {
		return _tttk_block_procid_while( blocked, msTimeOut );
	} else {
		return _tttk_block_app_while( app2run, blocked, msTimeOut );
	}
}
